File: backend/app/api/v1/skill_evolution.py
"""Self-evolving skills — admin-only debug/trigger surface. Dataset-building
and (later) optimization logic lives in backend/skill_evolution/, not here;
this file only wires HTTP in and out."""
from __future__ import annotations

import json
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.core import redis as redis_core
from app.core.guards import require_super_admin
from app.db.base import get_db
from app.db.models.user import User
from app.services import memory_service
from skill_evolution.dataset import build_dataset

router = APIRouter(prefix="/skill-evolution", tags=["skill-evolution"])


class DatasetExampleOut(BaseModel):
    query: str
    skill_content_snapshot: str
    output_trace: str | None
    label: str | None
    source: str


class DatasetPreviewOut(BaseModel):
    skill_id: uuid.UUID
    skill_name: str
    examples: list[DatasetExampleOut]
    summary: dict


@router.get("/skills/{skill_id}/preview-dataset", response_model=DatasetPreviewOut)
async def preview_dataset(
    skill_id: uuid.UUID,
    user: User = Depends(require_super_admin()),
    db: AsyncSession = Depends(get_db),
) -> DatasetPreviewOut:
    """Read-only: shows what an evolution run would build its eval dataset
    from, without triggering any optimization. No synthetic_generator is
    wired yet (that lands with the DSPy/GEPA optimizer), so a skill with too
    few real firings just previews a smaller, real-only dataset."""
    skill = await memory_service.get_skill(db, skill_id)
    if skill is None:
        raise HTTPException(status_code=404, detail="技能不存在")
    examples, summary = await build_dataset(db, skill)
    return DatasetPreviewOut(
        skill_id=skill.id,
        skill_name=skill.name,
        examples=[DatasetExampleOut(**vars(e)) for e in examples],
        summary=summary.to_dict(),
    )


class EvolveStatusOut(BaseModel):
    status: str  # idle | running | done | error
    detail: str | None = None
    finished_at: str | None = None


@router.post("/skills/{skill_id}/evolve", status_code=202)
async def trigger_evolve(
    skill_id: uuid.UUID,
    user: User = Depends(require_super_admin()),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Trigger an evolution run for one skill. Only ever produces a pending
    SkillProposal (Stage E's approval endpoint is the sole path that can
    change the live skill's content) — this just queues the optimizer."""
    skill = await memory_service.get_skill(db, skill_id)
    if skill is None:
        raise HTTPException(status_code=404, detail="技能不存在")

    r = redis_core.get_redis()
    status_key = redis_core.skill_evolution_status_key(str(skill_id))
    status_payload = json.dumps(
        {"status": "running", "started_at": datetime.now(timezone.utc).isoformat()},
        ensure_ascii=False,
    )
    # SET NX doubles as the run lock, same trick as memory_consolidate.
    ok = await r.set(status_key, status_payload, nx=True, ex=settings.skill_evolution_lock_ttl)
    if not ok:
        existing = await r.get(status_key)
        try:
            if existing and json.loads(existing).get("status") == "running":
                raise HTTPException(status_code=409, detail="该技能已有一次演化正在进行中")
        except (ValueError, TypeError):
            pass
        await r.set(status_key, status_payload, ex=settings.skill_evolution_lock_ttl)

    await redis_core.enqueue_prompt({"type": "skill_evolution", "skill_id": str(skill_id)})
    return {"status": "queued"}


@router.get("/skills/{skill_id}/evolve/status", response_model=EvolveStatusOut)
async def evolve_status(
    skill_id: uuid.UUID,
    user: User = Depends(require_super_admin()),
) -> EvolveStatusOut:
    r = redis_core.get_redis()
    raw = await r.get(redis_core.skill_evolution_status_key(str(skill_id)))
    data: dict = {}
    if raw:
        try:
            data = json.loads(raw)
        except (ValueError, TypeError):
            data = {}
    return EvolveStatusOut(
        status=data.get("status", "idle"),
        detail=data.get("detail"),
        finished_at=data.get("finished_at"),
    )

File: frontend/src/api/skillEvolution.ts
import { http } from "@/api/client";

export interface AdminSkill {
  id: string;
  name: string;
  description: string;
  content: string;
  owner_id: string | null;
  team_id: string | null;
  profile_id: string | null;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface EvolveStatus {
  status: "idle" | "running" | "done" | "error";
  detail?: string | null;
  finished_at?: string | null;
}

export interface DatasetExample {
  query: string;
  skill_content_snapshot: string;
  output_trace: string | null;
  label: string | null;
  source: "real" | "synthetic";
}

export interface DatasetPreview {
  skill_id: string;
  skill_name: string;
  examples: DatasetExample[];
  summary: { real_count?: number; synthetic_count?: number; earliest?: string | null; latest?: string | null };
}

export interface SkillProposal {
  id: string;
  skill_id: string;
  proposed_content: string;
  proposed_description: string | null;
  rationale: string | null;
  eval_score_before: number;
  eval_score_after: number;
  diff_ratio: number;
  dataset_summary: { real_count?: number; synthetic_count?: number; earliest?: string | null; latest?: string | null };
  status: "pending" | "approved" | "rejected";
  reviewed_by: string | null;
  reviewed_at: string | null;
  review_note: string | null;
  created_at: string;
}

export const skillEvolutionApi = {
  listSkills: (): Promise<AdminSkill[]> =>
    http.get("/skill-evolution/skills").then((r) => r.data),
  previewDataset: (skillId: string): Promise<DatasetPreview> =>
    http.get(`/skill-evolution/skills/${skillId}/preview-dataset`).then((r) => r.data),
  evolve: (skillId: string): Promise<{ status: string }> =>
    http.post(`/skill-evolution/skills/${skillId}/evolve`).then((r) => r.data),
  evolveStatus: (skillId: string): Promise<EvolveStatus> =>
    http.get(`/skill-evolution/skills/${skillId}/evolve/status`).then((r) => r.data),
  listProposals: (params?: { status?: string; skill_id?: string }): Promise<SkillProposal[]> =>
    http.get("/skill-evolution/proposals", { params }).then((r) => r.data),
  review: (proposalId: string, payload: { status: "approved" | "rejected"; review_note?: string }): Promise<SkillProposal> =>
    http.patch(`/skill-evolution/proposals/${proposalId}`, payload).then((r) => r.data),
};
